bip-0331.mediawiki

```diff
@@ -0,0 +1,428 @@
+information with one another, resulting in a more efficient and reliable way to propagate packages.
+
+(2) Eliminate the need for txid-based transaction relay and make orphan handling more robust.
+
+Txid-based transaction relay is problematic since a transaction's witness may be malleated without
+changing its txid; a node cannot use txid to deduplicate transactions it has already downloaded
+or validated. Ideally, two nodes that both support BIP339 wtxid-based transaction relay shouldn't
+ever need to use txid-based transaction relay.
+
+A single use case of txid-based relay remains: handling "orphan" transactions that spend output(s)
+from an unconfirmed transaction the receiving node is unaware of. Orphan transactions are very
+common for new nodes that have just completed Initial Block Download and do not have an up-to-date
+mempool. Nodes also download transactions from multiple peers. If the peer from which a child
+transaction was requested responds faster than the peer from which its parent was requested, that
+child is seen as an orphan transaction.
+
+Nodes may handle orphans by storing them in a cache and requesting any missing parent(s) by txid
+(prevouts specify txid, not wtxid). These parents may end up being orphans as well, if they also
+spend unconfirmed inputs that the node is unaware of. This method of handling orphans is problematic
+for two reasons: it requires nodes to allocate memory for unvalidated data received on the p2p
+network and it relies on txid-based relay between two wtxid-relay peers.
+
+This proposal makes orphan-fetching more efficient and no longer require txid-based relay.
+
+==Definitions==
+
+Given any two transactions Tx0 and Tx1 where Tx1 spends an output of Tx0, Tx0 is a '''parent''' of
+Tx1 and Tx1 is a '''child''' of Tx0.
+
+A transaction's '''ancestors''' include, recursively, its parents, the parents of its parents, etc.
+A transaction's '''descendants''' include, recursively, its children, the children of its children,
+etc. A transaction's parent is its ancestor, but an ancestor is not necessarily a parent.
+
+A '''package''' is a list of transactions, representable by a connected Directed Acyclic
+Graph (a directed edge exists between a transaction that spends the output of another transaction).
+In this proposal, a package is limited to unconfirmed transactions.
+
+An '''ancestor package''' consists of an unconfirmed transaction with all of its unconfirmed
+ancestors.
+
+In a '''topologically sorted''' package, each parent appears somewhere in the list before its child.
+
+==Specification==
+
+Ancestor Package Relay includes two parts: package information and transaction data download.
+A package information round is used to help a receiver learn what transactions are in a package and
+whether they want to download them. The transaction data round is used to help a node download
+multiple transactions in one message instead of as separate messages.
```

Comment:
I don't think it's that simple - we have cases of
- I downloaded this tx and rejected it for too low feerate, but I'm okay with trying again because it's with a package of other transactions I haven't seen before
- I haven't seen this tx before but there's another tx in the package that I know is consensus-invalid so I'm not going to download any of them

Reply:
> In future package types where you might get additional information, e.g., package feerates or similar this may make more sense to say.

Okay, "the information round *described here* doesn't really help you learn..."

> I haven't seen this tx before but there's another tx in the package that I know is consensus-invalid so I'm not going to download any of them

In that case, (on the implementation side) would we need different reject caches, one for "consensus invalid", one for "policy invalid"? Not sure that would be worth the hassle (honest peers sending consensus invalid stuff should be extremely rare; attackers could just send novel txs each time avoiding the cache)

Reply:
> In that case, (on the implementation side) would we need different reject caches, one for "consensus invalid", one for "policy invalid"?

Yes. Two rejection caches is necessary, see https://github.com/glozow/bitcoin/pull/8#issuecomment-1453791776. But not consensus vs policy, just ones that are eligible for reconsideration in a package (i.e. too low feerate or missing ephemeral anchor spend) vs ones that are not. I said "consensus invalid" to provide a clear example, but I just mean "something we don't need to validate again, even if it's in a package." If something was larger than max standard size or has too many ancestors, trying it in a package won't make it valid.

Reply:
I guess as long as they're global caches and not per-peer it doesn't matter much anyway.

Reply:
I'm going to resolve this for now, but let me know if you still feel that I should change this. Right now the text is "The package information round is used to help a receiver learn what transactions are in a package and decide whether they want to download them."